fenics_ice/inout.py

```diff
@@ -306,37 +306,85 @@ def field_from_vel_file(infile, field_name):
 
     infile = h5py.File(infile, 'r')
+    keys = list(infile.keys())
+
+    assert 'u_obs' in keys, \
+        f"Your vel file has not been configure correctly " \
+        f"make sure you have velocity data with the correct var names " \
+        f"x, y, u_obs, v_obs, u_obs_std, v_obs_std "
 
-    x_obs = field_from_vel_file(infile, 'x')
-    y_obs = field_from_vel_file(infile, 'y')
-    u_obs = field_from_vel_file(infile, 'u_obs')
-    v_obs = field_from_vel_file(infile, 'v_obs')
-    u_std = field_from_vel_file(infile, 'u_std')
-    v_std = field_from_vel_file(infile, 'v_std')
+    # Read composite mean of velocity data as default
     mask_vel = field_from_vel_file(infile, 'mask_vel')
+    x_comp = field_from_vel_file(infile, 'x')
+    y_comp = field_from_vel_file(infile, 'y')
+    u_comp = field_from_vel_file(infile, 'u_obs')
+    v_comp = field_from_vel_file(infile, 'v_obs')
+    u_comp_std = field_from_vel_file(infile, 'u_std')
+    v_comp_std = field_from_vel_file(infile, 'v_std')
+
+    if params.inversion.use_cloud_point_velocities:
+        logging.warning(f"You are using cloud point data for optimizing J, "
+                        f"inversion results might not be smooth")
+        assert 'u_cloud' in keys, \
+            f"Your vel file has not been configure correctly " \
+            f"make sure you have cloud point data with the correct var names " \
+            f"x_cloud, y_cloud, u_cloud, v_cloud, u_cloud_std, v_cloud_std"
+        # Read cloud point observations to be used in the Inversion and
+        # Cost function optimization only!
+        x_obs = field_from_vel_file(infile, 'x_cloud')
+        y_obs = field_from_vel_file(infile, 'y_cloud')
+        u_obs = field_from_vel_file(infile, 'u_cloud')
+        v_obs = field_from_vel_file(infile, 'v_cloud')
+        u_std = field_from_vel_file(infile, 'u_cloud_std')
+        v_std = field_from_vel_file(infile, 'v_cloud_std')
+    else:
+        x_obs = x_comp
+        y_obs = y_comp
+        u_obs = u_comp
+        v_obs = v_comp
+        u_std = u_comp_std
+        v_std = v_comp_std
 
     assert x_obs.size == y_obs.size == u_obs.size == v_obs.size
-    assert v_obs.size == u_std.size == v_std.size == mask_vel.size
+    assert v_obs.size == u_std.size == v_std.size
+    assert x_comp.size == y_comp.size == u_comp.size == v_comp.size
+    assert v_comp.size == u_comp_std.size == v_comp_std.size == mask_vel.size
 
     uv_obs_pts = np.vstack((x_obs, y_obs)).T
+    uv_comp_pts = np.vstack((x_comp, y_comp)).T
 
     if model is not None:
         model.uv_obs_pts = uv_obs_pts
         model.u_obs = u_obs
         model.v_obs = v_obs
         model.u_std = u_std
         model.v_std = v_std
+        model.uv_comp_pts = uv_comp_pts
+        model.u_comp = u_comp
+        model.v_comp = v_comp
+        model.u_comp_std = u_comp_std
+        model.v_comp_std = v_comp_std
         model.mask_vel = mask_vel
     else:
```

Comment:
No `else` here? i.e. if `model` is supplied then store the data on `model`, and also return the data. Even better the data could be stored in a map on `model`, e.g. `self.u_data = inout.read_vel_obs([...])`. This avoids setting too many `model` attributes.

Reply:
yes I like this it would be better also will be better to also return a directory instead of all this long list of variables